nfs-ganesha: Build with Ceph and DBus support

dsg22 opened the pull request:
This enables using the Ceph FSAL to serve files from CephFS. Also enable DBus.

Testing was done against nixpkgs-25.05 using callPackage due to breakage on unstable, but I don't think it matters here.

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

markuskowa commented on pkgs/by-name/nf/nfs-ganesha/package.nix:
```diff
@@ -16,6 +16,7 @@
   flex,
   nfs-utils,
   acl,
+  ceph,
```
```suggestion
  useCeph ? false,
  ceph,
```
Let's make it conditional. This is a backend that pulls in a lot of dependencies.

markuskowa reviewed (changes requested):
ceph is currently broken on master (i.e. arrow-cpp is actually broken).

dsg22 commented:
Yeah, I noticed the ceph build breakage. I tested this by using the module from master on 25.05. Sorry, should have mentioned that.

It turns out my project won't end up using nfs-ganesha after all, so I don't have a need for this functionality anymore. I'll still try to update the PR to make the ceph dependency optional when I have some free time. Thanks for taking the time to look at it.

dsg22 committed:
nfs-ganesha: Add (default disabled) Ceph support

CephFS and RGW FSAL, storage of recovery info in RADOS objects, and RADOS
URL support.

dsg22 committed:
nfs-ganesha: Add DBus support for metrics and management

https://github.com/nfs-ganesha/nfs-ganesha/wiki/Dbusinterface

dsg22 commented:
Turns out I do need nfs-ganesha after all :)

I updated the Ceph changes to make them optional and also updated to build other Ceph features like the rados_ng recovery backend.

Since I was working on this, I also made two unrelated changes - one is to bump the nfs-ganesha version to the new 7.0 release, and the other is to enable DBus (can be used for runtime management and to collect service metrics). I figured it would be easier to have all three in the same PR, but if you prefer I can open separate PRs for those. They are all self-contained in their own commits.

The build patch for DBus is a bit ugly, but I just could not get cmake to pick up the DBus headers and libs through pkg-config.

dsg22 commented:
I pulled the v7.0 patch, I'll submit that as a separate PR. Sorry for the churn.

dsg22 commented:
Okay, the v7.0 patch is ready. Needed to upgrade ntirpc as well. I added it to this PR, but let me know if you prefer me to split them up.

I also tested this on nixpkgs master, since the build failures seem to be fixed now. NFS serving works, no problems observed.

markuskowa commented:
The patches have subject headers. Did you create them locally, or did you download them from some other distro or pull request?

markuskowa reviewed (commented):
I think otherwise it looks good now. I will test it later tonight to make sure nothing broke.

dsg22 commented:
> I think otherwise it looks good now. I will test it later tonight to make sure nothing broke.

Hold off on that, after rebasing on new master I'm seeing a build failure. I'll push an update once I've dealt with that.

markuskowa reviewed (commented):
I tested it, and it looks good now. Everything still works (and now with DBUS, which is nice).
The last comments are just on nixpkgs coding convention. Once these two are integrated, we can merge the PR.

dsg22 committed:
nfs-ganesha: Coding style updates from review

Co-authored-by: Markus Kowalewski <markus.kowalewski@gmail.com>

dsg22 commented:
Sure, added the requested changes.
